Fixed stroke width, default miter, matrix parsing, added testcase

jcward opened the pull request:
See individual commits:
- Default joint was round, fixed to miter per: https://www.w3.org/TR/SVG/painting.html#StrokeProperty
- Fixed matrix regex, was greedy matching the floats, creating matrices with NaN
- Fixed the sqrt(2) factor in line width scale

Created a testcase which exercises all of the above:

![image](https://cloud.githubusercontent.com/assets/2192439/14920769/40186768-0dec-11e6-8959-1b707cdd2417.png)

```
<svg id="svg1" viewBox="0 0 256 256" xmlns="http://www.w3.org/2000/svg" width="256" height="256">
  <rect width="256" height="256" style="fill:#FFF" />
  <rect x="10" y="10" width="100" height="20" style="fill:#f00;stroke:#000;stroke-width:10;" transform="scale(2.0)"/>
  <rect x="30" y="25" width="40" height="20" style="fill:#0f0;stroke:#000;stroke-width:15;" transform="matrix(3.38073983743877, 0.905868727550796, -0.905868727550796, 3.38073983743877, 0, 0)"/>
</svg>
```

ashes999 commented:
@ibilon @jgranick can you please review this?

jcward commented:
I realize I've done the default joint style slightly wrong for inheritance, but it won't matter until later when the properties are properly inherited. I've said that "not defining a joint style == define as miter" -- but later when they're inherited, inheritance needs to take precedence over this non-definition.

But since property inheritance is currently non-existent, and this is an improvement on the current state, I might take this PR anyway. We're taking baby steps in the right direction. ;)

ashes999 commented:
Please review @ibilon

ibilon commented on src/format/svg/SVGRenderer.hx:
```diff
@@ -29,6 +29,7 @@ typedef ObjectFilter = String->GroupPath->Bool;
 
 class SVGRenderer
 {
+    public static inline var SQRT2:Float = 1.41421356237;
```
Since it'd only be computed once I'd go with `Math.sqrt(2)`.

jcward replied:
Sounds good.

jcward replied:
Ah, I was just following this precedent: https://github.com/openfl/svg/blob/master/src/format/svg/SVGData.hx#L29-L30

But whatever, happy to use `Math.sqrt(2)`

ibilon replied:
Ah, `Math.sqrt(2)` isn't considered constant, so can't be in a `static inline` :/

ibilon commented on src/format/svg/SVGData.hx:
```diff
@@ -581,9 +581,9 @@ class SVGData extends Group {
 		path.stroke_colour = getStrokeStyle ("stroke", inPath, styles, null);
 		path.stroke_width = getFloatStyle ("stroke-width", inPath, styles, 1.0);
-		path.stroke_caps = getStyleAndConvert ("stroke-linecap", inPath, styles, null, 
+		path.stroke_caps = getStyleAndConvert ("stroke-linecap", inPath, styles, CapsStyle.NONE, 
```
https://www.w3.org/TR/SVG/painting.html#StrokeLinecapProperty seems to indicate that `butt` is the default value

jcward replied:
Ah, there're two properties, cap style and joint style. The line caps default to butt as you pointed out, but the line joint style is miter: https://www.w3.org/TR/SVG/painting.html#StrokeLinejoinProperty)

ibilon commented:
@jcward do you have an example of matrix string that failed with the previous regex?

Also this seems to have conflict now.

jcward commented:
I'll see if I can push an update in the next hour or so.

jcward commented:
@ibilon ok, I've fixed the merge conflict and added a simple matrix rotation testcase that will coredump before this matrix regexp fix is in place:

```
<svg width="100" height="100" xmlns="http://www.w3.org/2000/svg" xmlns:svg="http://www.w3.org/2000/svg">
 <g>
  <title>Layer 1</title>
  <rect fill="#ffffff" x="0" y="0" width="100" height="100" />
  <rect fill="#ff0000" stroke="#000000" stroke-width="0" stroke-dasharray="null" stroke-linejoin="null" stroke-linecap="null" x="-25" y="50" width="50" height="50" transform="matrix(0.7071, -0.7071, 0.7071, 0.7071, 0, 0)" />
 </g>
</svg>
```

Results in:
```
>haxe test.hxml 

MUnit Results
------------------------------
Class: SvgGenerationTest ....x.Segmentation fault (core dumped)
Error: Command failed with error 139
```

Because the old regex captures parses nulls values into the Matrix it creates.

TheSHEEEP commented:
Any specific reason not to apply this PR?

ashes999 commented:
@TheSHEEEP  no reason in particular.

@ibilon does this PR meet your approval? I can merge it if so.

ibilon commented:
@ashes999 Sorry forgot about it, all good for me.
